### ISSUE TITLE
Creating React Components as Posts - Frontmatter, Templating, and More

### ISSUE THREAD

**Issue by frewinchristopher** [gatsbyjs/gatsby#9862]
## Summary

Is there a standard. I've been searching for a few hours now, [I've found this from gatsby-starter-blog](https://github.com/gatsbyjs/gatsby-starter-blog/commit/e333cb1aadbba2d5053a3b471fa841b29b3187c2) from @KyleAMathews  for v1, but this was an old commit and `pages/index.js`
is now written in a completely different fashion for v2, not to mention the fact that the `exports.data` syntax is now forbidden in v2 when using `import`

I also found [this issue in gatsby-starter-blog](https://github.com/gatsbyjs/gatsby-starter-blog/issues/61)... but I'm the only person who has replied so far.

So, my current workaround is the following, using the `gatsby-transformer-javascript-frontmatter` plugin - but that only helps with frontmatter. I still need to create my own excerpt for component type posts, here is an example JS post to show what I mean:

```
const React = require('react')

class TestJSPage extends React.Component {
  render () {
    return (
      <div>
        <h1>Here's a Test</h1>
        <p>Here's some content for my javascript style post. I'll have to write the excerpt manually because I dont know how to do it automatically with JS posts.</p>
      </div>
    );
  }
}

export default TestJSPage;

// front matter, javascript style
export const frontmatter = {
  title: "Here's a Test",
  excerpt: "Here's some content for my javascript style...",
  date: "2018-10-12"
}
```

At least in that way in `pages/index.js` I can access the 'excerpt' as a frontmatter- level component of `node` (reverting to the original way if `excerpt` is at the node level:

```
{ node.frontmatter.excerpt && <p dangerouslySetInnerHTML={{ __html: node.frontmatter.excerpt }} /> }
{ node.excerpt && <p dangerouslySetInnerHTML={{ __html: node.excerpt }} /> }
```

I then needed to copy all the `allMarkdownRemark` in both `pages/index.js` and `gatsby-node.js` - but that gets me only halfway.

I would like to also  incorporate .js posts into the graphql query in the `blog-post.js` template file: 

```
markdownRemark(fields: { slug: { eq: $slug } }) {
      id
      excerpt
      html
      frontmatter {
        title
        date(formatString: "MMMM DD, YYYY")
      }
    }
```

Since we presumably don't have any id, excerpt, or html from component type posts.

Until I figure that out, I am hardcoding in each .js component post the site title (linking back to the main page), as well as the post title itself and date. They are NOT passed down automatically as they are with .md posts.

### File contents 

`gatsby-config.js`: as stated, added `gatsby-transformer-javascript-frontmatter`
`gatsby-node.js`: as stated, added near copy of `allMarkdownRemark` query, just for the package `gatsby-transformer-javascript-frontmatter`:

```
allJavascriptFrontmatter {
              edges {
                node {
                  fields {
                    slug
                  }
                  frontmatter {
                    date(formatString: "DD MMMM, YYYY")
                    title
                  }
                }
              }
            }
```

I then concat them together:

```
// Create blog posts pages.
let posts = result.data.allMarkdownRemark.edges;
posts.concat(result.data.allJavascriptFrontmatter.edges);
```

Basically, I'm looking for a best-practices way to do this, not the hacky way that I've put together. Is this anywhere in the docs that I've missed? I realize .js can be put in the root of `/pages`, but I want to organize posts like the `.md` are: each in their own folder.
